daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala

```diff
@@ -274,10 +284,19 @@ object TransactionCoder {
           children <- childrenOrError
           cv <- decodeVal(protoExe.getChosenValue)
           templateId <- ValueCoder.decodeIdentifier(protoExe.getTemplateId)
-          controllers <- toPartySet(protoExe.getControllersList)
+          actingParties <- toPartySet(protoExe.getActorsList)
+          encodedControllers <- toPartySet(protoExe.getControllersList)
+          controllers <- if (!(txVersion precedes minNoControllers)) {
+            if (encodedControllers.isEmpty) {
+              Right(actingParties)
+            } else {
+              Left(DecodeError(s"As of version $txVersion, exercise controllers must be empty."))
+            }
+          } else {
+            Right(encodedControllers)
```

Comment:
@bitonic As you say,

> we can't [delete the controllers field] if we keep the validation separate.

I want to remove the possibility that the field is used for anything but that separate validation, then, rather than leaving the separate `controllers` field open to "interpretation", so to speak; depending on how we report validation failures here, that can be as simple as

```suggestion
            Right(decodedControllers != actingParties)
```

and the other case on line 291 is `Right(false)`.